feat(api): add resource group overview endpoint and filtering

AdriiiPRodri opened the pull request:
### Context

This PR adds a new Resource Group API feature that enables grouping and filtering findings by logical resource categories (storage, compute, security, database, etc.). This provides users with a high-level view of their security posture organized by resource type, complementing the existing categories functionality.

Related to the Resource Inventory component requirements.

### Description

Add support for resource group aggregations to enable grouping findings by logical resource categories. This feature mirrors the existing categories implementation pattern.

**Changes:**

**API:**
- Add `ScanResourceGroupSummary` model for pre-aggregated metrics per scan
- Add `group` field to `Finding`, `Resource`, and `ScanResourceGroupSummary` models
- Add `/api/v1/overviews/groups` endpoint with provider filtering
- Add `group` and `group__in` filters to findings and resources endpoints
- Include `groups` in metadata responses
- Add `aggregate_resource_group_counts` for real-time scan aggregation
- Add `backfill_scan_resource_group_summaries` task for historical data
- Add `ResourceGroupOverviewFilter` and `ResourceGroupOverviewSerializer`

The endpoint returns aggregated counts (`total_findings`, `failed_findings`, `new_failed_findings`, `resources_count`) grouped by `group` and `severity`, following the same pattern as the categories endpoint.

**Field naming convention (consistent with existing API patterns):**
| Model Field (singular) | Metadata (plural) | Filter |
|------------------------|-------------------|--------|
| `group` | `groups` | `group`, `group__in` |

### Manual Testing Performed

All endpoints verified working:
- ✅ `GET /api/v1/findings` → returns `group` field
- ✅ `GET /api/v1/resources` → returns `group` field  
- ✅ `GET /api/v1/findings?filter[group]=monitoring` → filter works
- ✅ `GET /api/v1/findings?filter[group__in]=monitoring,security` → multi-filter works
- ✅ `GET /api/v1/overviews/groups` → returns group aggregations
- ✅ `GET /api/v1/findings/metadata` → returns `groups` (plural)

All 1426 API tests pass.

### Steps to review

1. Review the new migration `0066_finding_resource_group_scanresourcegroupsummary.py`
2. Test the new endpoint:
   - `GET /api/v1/overviews/groups` - returns all resource groups
   - `GET /api/v1/overviews/groups?filter[group]=storage` - filter by specific group
   - `GET /api/v1/overviews/groups?filter[provider_type]=aws` - filter by provider
3. Test findings filter:
   - `GET /api/v1/findings?filter[group]=storage`
   - `GET /api/v1/findings?filter[group__in]=storage,security`
4. Verify metadata endpoint includes `groups`:
   - `GET /api/v1/findings/metadata`
   - `GET /api/v1/resources/metadata`
5. Run a scan and verify `ScanResourceGroupSummary` records are created
6. Verify `resources_count` correctly counts unique resources (not total findings)

### Checklist

- Are there new checks included in this PR? **No**
- [x] Review if the code is being covered by tests.
- [ ] Review if code is being documented following this specification
- [ ] Review if backport is needed.
- [ ] Review if is needed to change the Readme.md
- [ ] Ensure new entries are added to CHANGELOG.md, if applicable.

#### UI
- N/A (API-only changes)

#### API
- [x] Verify if API specs need to be regenerated.
- [x] Check if version updates are required (e.g., specs, Poetry, etc.).
- [ ] Ensure new entries are added to CHANGELOG.md, if applicable.

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

AdriiiPRodri committed:
docs: add ResourceGroup field specification to developer guide

AdriiiPRodri committed:
chore: rename identity to IAM

AdriiiPRodri committed:
feat(api): add resource group overview endpoint and filtering

AdriiiPRodri committed:
test(api): add unit tests

AdriiiPRodri committed:
chore: update changelog

AdriiiPRodri committed:
chore: restore v1

AdriiiPRodri committed:
refactor(api): remove fallback

jfagoagas commented on api/CHANGELOG.md:
```diff
@@ -6,6 +6,8 @@ All notable changes to the **Prowler API** are documented in this file.
 
 ### Added
 - Support AlibabaCloud provider [(#9485)](https://github.com/prowler-cloud/prowler/pull/9485)
+- New endpoint to retrieve an overview of the resource groups based on finding severities [(#9694)](https://github.com/prowler-cloud/prowler/pull/9694)
```
```suggestion
- `/api/v1/overviews/resource_groups` to retrieve an overview of the resource groups based on finding severities [(#9694)](https://github.com/prowler-cloud/prowler/pull/9694)
```

jfagoagas commented on api/src/backend/api/migrations/0066_finding_resource_group_scanresourcegroupsummary.py:
```diff
@@ -0,0 +1,127 @@
+        migrations.AddField(
+            model_name="finding",
+            name="resource_group",
+            field=models.CharField(
```
I'd add a `TextField` just in case. I know 50 chars should be more than enough but I prefer using that one.

AdriiiPRodri replied:
Done

jfagoagas commented on api/src/backend/tasks/jobs/backfill.py:
```diff
@@ -341,3 +342,78 @@ def backfill_scan_category_summaries(tenant_id: str, scan_id: str):
     return {"status": "backfilled", "categories_count": len(category_counts)}
+
+
+def backfill_scan_resource_group_summaries(tenant_id: str, scan_id: str):
```
Why do we need this backfill? I thought this feature will populate data once released as the new `ResourceGroup` was not present before.

jfagoagas reviewed (changes requested):
That's great @AdriiiPRodri I did a first pass without a functional test and left some comments/suggestions. Thanks!

AdriiiPRodri commented on api/src/backend/api/migrations/0066_finding_resource_group_scanresourcegroupsummary.py:
```diff
@@ -0,0 +1,127 @@
+            ),
+        ),
+        migrations.CreateModel(
+            name="ScanResourceGroupSummary",
```
In the current code, resource groups are tied to findings (derived from `check_metadata.resourcegroup`), so they end up being associated with the scan that created those findings. `ScanResourceGroupSummary` is also per scan (and tenant), not per provider directly, provider is inferred via the scan (`scan__provider`)

AdriiiPRodri committed:
refactor(api): change to TextField

AdriiiPRodri commented on api/CHANGELOG.md:
```diff
@@ -6,6 +6,8 @@ All notable changes to the **Prowler API** are documented in this file.
 ### Added
 - Support AlibabaCloud provider [(#9485)](https://github.com/prowler-cloud/prowler/pull/9485)
+- New endpoint to retrieve an overview of the resource groups based on finding severities [(#9694)](https://github.com/prowler-cloud/prowler/pull/9694)
+- Endpoints `GET /findings` and `GET /findings/metadata/latest` now support the resource_group filter [(#9694)](https://github.com/prowler-cloud/prowler/pull/9694)
```
Done

AdriiiPRodri committed:
chore: remove duplicated entry

AdriiiPRodri commented on docs/developer-guide/checks.mdx:
```diff
@@ -341,6 +341,31 @@ A high-level classification that groups checks by the type of cloud resource the
 
 The group is determined by the resource type being audited, not the service. For example, an EC2 security group check would use `network` (not `compute`), while an EC2 instance check would use `compute`.
 
+#### ResourceGroup
```
Should happen in the rebase, fixed thanks!

Alan-TheGentleman committed:
feat(api): add resource_group field to Resource model

- Add resource_group field to Resource model (0067 migration)
- Add backfill migration to populate from findings (0068 migration)
- Add resource_group filter (exact, in) to ResourceFilter
- Add resource_group to ResourceSerializer
- Update API spec with resource_group field and filter params
- Populate resource_group from check metadata in scan job
- Add tests for resource_group filter and field presence in response

Alan-TheGentleman committed:
refactor(api): remove resource_group backfill migration

Remove migration 0068 that backfills resource_group from findings.
This operation is too expensive for production databases.

The resource_group field will only be populated for new resources
during scans. Existing resources will have NULL until re-scanned.

Alan-TheGentleman committed:
feat(api): add resource_groups to resources metadata endpoints

- Add resource_groups field to ResourceMetadataSerializer
- Add resource_groups query to /resources/metadata endpoint
- Add resource_groups query to /resources/metadata/latest endpoint
- Add resource_group filter to LatestResourceFilter

jfagoagas reviewed (changes requested):
After our conversation:
- Rename `resource_group` to `group`
- Mark `resource_type` as deprecated in the API spec

Alan-TheGentleman committed:
refactor(api): rename resource_group to group in Resource model

- Rename resource_group field to group in Resource model
- Update migration 0067 with new field name
- Update ResourceSerializer and ResourceMetadataSerializer
- Update filters (ResourceFilter, LatestResourceFilter)
- Update views metadata queries
- Update scan.py Resource creation/update logic
- Update fixtures and tests
- Update API spec with new field names and filters
- Mark resource type field as deprecated in API spec
- Add groups assertions to resources metadata tests

AdriiiPRodri reviewed (commented):
I think there's a bit of confusion between group and groups, at least for now. That said, I don't think this should change: resources should belong to a single group only

jfagoagas commented:
> I think there's a bit of confusion between group and groups, at least for now. That said, I don't think this should change: resources should belong to a single group only

That's correct. The relationship is 1:1.

Alan-TheGentleman committed:
refactor(api): rename resource_group to group across Finding and Summary models

Rename the field from 'resource_group' to 'group' for consistency with the
API naming conventions (singular for model fields, plural for metadata).

Changes:
- Models: Finding.group, ScanResourceGroupSummary.group (migration updated)
- Filters: filter[group], filter[group__in]
- Serializers: 'group' field in responses
- Views: /api/v1/overviews/groups endpoint (was /resource_groups)
- Metadata: 'groups' (plural) in metadata responses
- Jobs: scan.py and backfill.py updated to use 'group'
- Tests: All fixtures and assertions updated

Manual testing performed:
- GET /api/v1/findings -> returns 'group' field ✓
- GET /api/v1/resources -> returns 'group' field ✓
- GET /api/v1/findings?filter[group]=monitoring -> works ✓
- GET /api/v1/findings?filter[group__in]=monitoring,security -> works ✓
- GET /api/v1/overviews/groups -> returns group aggregations ✓
- GET /api/v1/findings/metadata -> returns 'groups' (plural) ✓

All 1426 API tests pass.

Note for existing deployments: If migration 0066 was already applied with
the old column name 'resource_group', run:
  ALTER TABLE findings RENAME COLUMN resource_group TO "group";
  ALTER TABLE scan_resource_group_summaries RENAME COLUMN resource_group TO "group";

AdriiiPRodri committed:
refactor(api): rename resource_group to group

vicferpoy committed:
Merge branch 'master' into PROWLER-37-resource-inventory-component-api

vicferpoy committed:
feat: add resource group model and fields

vicferpoy committed:
feat: add resource group overview endpoint and filters

vicferpoy committed:
feat: implement resource group aggregation in scan and backfill

vicferpoy committed:
test: add resource group feature tests

vicferpoy committed:
docs: update API spec and changelog for resource groups

vicferpoy committed:
style: apply ruff

vicferpoy committed:
fix: use kebab-case for response objects

jfagoagas reviewed (approved):
Thanks for the final changes @vicferpoy, as always:
<img src="https://media1.giphy.com/media/v1.Y2lkPWJkM2VhNTdlYXdibWlkMnQxMGp5YmRvaXBqeGx3NW81NDYwYnR4eHpybDVjMHpydSZlcD12MV9naWZzX3NlYXJjaCZjdD1n/3q3QK6KyDVUBzih7hB/giphy.gif"/>
